server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +412,300 @@ public int hashCode() {
+        private void updateCleanupKeyToCountMapOnCacheInsertion(CleanupKey cleanupKey) {
+            if (stalenessThreshold == 0.0) {
+                return;
+            }
+            IndexShard indexShard = (IndexShard) cleanupKey.entity.getCacheIdentity();
+            if (indexShard == null) {
+                logger.warn("IndexShard is null for CleanupKey: {} while cleaning Indices Request Cache", cleanupKey.readerCacheKeyId);
+                return;
+            }
+            ShardId shardId = indexShard.shardId();
+
+            // If the key doesn't exist, it's added with a value of 1.
+            // If the key exists, its value is incremented by 1.
+            cleanupKeyToCountMap.computeIfAbsent(shardId, k -> ConcurrentCollections.newConcurrentMap())
+                .merge(cleanupKey.readerCacheKeyId, 1, Integer::sum);
+        }
+
+        private synchronized void updateCleanupKeyToCountMapOnCacheEviction(CleanupKey cleanupKey) {
+            if (stalenessThreshold == 0.0) {
+                return;
+            }
+            IndexShard indexShard = (IndexShard) cleanupKey.entity.getCacheIdentity();
+            if (indexShard == null) {
+                logger.warn("IndexShard is null for CleanupKey: {} while cleaning Indices Request Cache", cleanupKey.readerCacheKeyId);
+                return;
             }
+            ShardId shardId = indexShard.shardId();
+            // If the key doesn't exist, ignore
+            ConcurrentMap<String, Integer> keyCountMap = cleanupKeyToCountMap.get(shardId);
+            if (keyCountMap != null) {
+                keyCountMap.compute(cleanupKey.readerCacheKeyId, (key, currentValue) -> {
+                    // If the key is not present, no action is needed, return null.
+                    if (currentValue == null) return null;
+                    // Calculate the new value.
+                    int newValue = currentValue - 1;
+                    // decrement the stale key count
+                    staleKeysCount.decrementAndGet();
+                    // Remove the key if the new value is zero by returning null; otherwise, update with the new value.
+                    return newValue == 0 ? null : newValue;
+                });
+            }
+        }
+
+        /**
+         * Updates the count of stale keys in the cache.
+         * This method is called when a CleanupKey is added to the keysToClean set.
+         *
+         * It increments the staleKeysCount by the count of the CleanupKey in the cleanupKeyToCountMap.
+         * If the CleanupKey's readerCacheKeyId is null or the CleanupKey's entity is not open, it increments the staleKeysCount
+         * by the total count of keys associated with the CleanupKey's ShardId in the cleanupKeyToCountMap and removes the ShardId from the map.
+         *
+         * @param cleanupKey the CleanupKey that has been marked for cleanup
+         */
+        private void updateStaleKeysCount(CleanupKey cleanupKey) {
```

Comment:
Better rename to something specific like "incrementStaleKeysCountOn...."?

Reply:
True going with `incrementStaleKeysCount`